docs/workshops/20220920-intro-to-git-and-github/lesson.md

```diff
@@ -0,0 +1,611 @@
+A **branch** is an independent line of development of files in a repository. 
+You can think of it like making a copy of a document and making changes to that copy only: the original stays the same, but the copy diverges with new content.
+In a branch, all files are copied and can be changed.
+
+
+<center>
+![](git_branch_collab.png){ width="600" }
+</center>
+
+Branches are particularly powerful for collaborating.
+They allow multiple people to work on the same code base and seamless integrate changes.
+
+By default, you start on the `main` branch.
+You can see which branch you're on locally using the `git branch` command.
+
+```
+git branch
+``` 
+
+Let's create a branch and make changes to it.
+We can create a new local branch using `git checkout -b`.
+`git checkout` tells git we want to switch which branch we're currently on, while the `-b` flag tells git that we're creating a new branch.
+
+```
+git checkout -b my-first-branch 
+```
+
+Now, if we run `git branch` we'll see we're on the branch `my-first-branch`.
+
+```
+git branch
+```
+
+To go back to the `main` branch, we can use the `git checkout` command without the flag `-b`.
+
+```
+git checkout main
+```
+
+
+Branch names can be used to convey information that can help you and others keep track of what you're doing.
+For example, you can prepend your branch names with your initials and a slash to make it clear that you were the one who created the branch, and you can specify a name that is associated with the file changes made on the branch: `<your initials>/<brief description of the code change>`. 
+Example: `ter/git-workshop`.
+
+Let's practice making a branch using these conventions. 
+In our branch, we'll update our `README.md` file.
+Use the code below to create a new branch, but remember to substitute out `your-initials` for your initials.
+
+```
+git checkout -b your-initials/update-readme.
+```
+
+Whenever you create a new branch, it branches off from the branch you're currently in when you make the new branch.
+In this case, we started in `main`, so our branch will branch off of from here.
+
+Next, let's add some changes to our `README.md` file and run `git status`.
+
+```
+echo "adding more text to the README file" >> README.md
+git status
+```
+
+Stage the changes with `git add`:
+
+```
+git add README.md
+```
+
+And commit them:
+```
+git commit -m "update readme w more text"
+```
+
+Run `git push`:
+
+```
+git push
+```
+
+This gives us an error message!
+
+```
+fatal: The current branch tmp has no upstream branch.
+To push the current branch and set the remote as upstream, use
+
+    git push --set-upstream origin ter/update-readme
+```
+
+While we created our branch locally, we haven't created the same branch remotely. 
+We need to tell `git` where to put our changes in the remote repository.
+Luckily, the error message points us to the code we need to run to set the remote branch.
+
+```
+git push --set-upstream origin ter/update-readme
```

Comment:
ooh tell me more! I know nothing about this setting! I think that is desirable -- not sure if it would best be delivered in this workshop vs. in a later AUG efficiency with git. We could always add it to this lesson material and I could skip covering it if we're short on time

Reply:
I think it's a config:
```
git config --global push.default current
git config --global remote.pushDefault origin
```